core/include/scipp/core/multi_index.h

```diff
@@ -310,24 +383,188 @@ template <scipp::index N> class MultiIndex {
+    std::fill(coord_it(), coord_end(), 0);
+    const auto last_dim = (bin_ndim() == 0 ? m_nested_dim_index : m_ndim - 1);
+    coord(last_dim) = shape(last_dim);
+
+    for (scipp::index data = 0; data < N; ++data) {
+      // Only one dim contributes, all others have coord = 0.
+      m_bin[data].m_bin_index = coord(last_dim) * stride(last_dim, data);
+      load_bin_params(data);
+    }
+  }
+
+  scipp::index flat_index(const scipp::index i_data, scipp::index begin_index,
+                          const scipp::index end_index) {
+    scipp::index res = 0;
+    for (; begin_index < end_index; ++begin_index) {
+      res += coord(begin_index) * stride(begin_index, i_data);
+    }
+    return res;
+  }
+
+  /// Stride for each operand in each dimension.
+  [[nodiscard]] scipp::index &stride(const scipp::index dim,
+                                     const scipp::index data) noexcept {
+    return m_buffer[data + dim * N];
+  }
+
+  [[nodiscard]] const scipp::index &
+  stride(const scipp::index dim, const scipp::index data) const noexcept {
+    return m_buffer[data + dim * N];
+  }
+
+  template <size_t... I, class... StridesArgs>
+  void store_strides_impl(const scipp::index ndim, const scipp::index first_dim,
+                          std::index_sequence<I...>,
+                          const StridesArgs &... strides) noexcept {
+    (
+        [&]() {
+          for (scipp::index dim = 0; dim < ndim; ++dim) {
+            stride(first_dim + dim, I) = strides[ndim - 1 - dim];
+          }
+        }(),
+        ...);
+  }
+
+  template <class... StridesArgs>
+  void store_strides(const scipp::index ndim, const scipp::index first_dim,
+                     const StridesArgs &... strides) noexcept {
+    store_strides_impl(ndim, first_dim,
+                       std::index_sequence_for<StridesArgs...>(), strides...);
+  }
+
+  /// Current index in iteration dimensions for both bin and inner dims.
+  [[nodiscard]] scipp::index &coord(const scipp::index dim) noexcept {
+    return *coord_it(dim);
+  }
+
+  [[nodiscard]] const scipp::index &
+  coord(const scipp::index dim) const noexcept {
+    return *coord_it(dim);
+  }
+
+  [[nodiscard]] auto coord_it(const scipp::index dim = 0) const noexcept {
+    return m_buffer.get() + (std::max(m_ndim, scipp::index{2}) * N + dim);
+  }
+
+  [[nodiscard]] auto coord_end() const noexcept {
+    return m_buffer.get() + (std::max(m_ndim, scipp::index{2}) * (N + 1));
+  }
+
+  /// Shape of the iteration dimensions for both bin and inner dims.
+  [[nodiscard]] scipp::index &shape(const scipp::index dim) noexcept {
+    return *shape_it(dim);
+  }
+
+  [[nodiscard]] const scipp::index &
+  shape(const scipp::index dim) const noexcept {
+    return *shape_it(dim);
+  }
+
+  [[nodiscard]] auto shape_it(const scipp::index dim = 0) const noexcept {
+    return m_buffer.get() + (std::max(m_ndim, scipp::index{2}) * (N + 1) + dim);
+  }
+
+  [[nodiscard]] auto shape_end() const noexcept {
```

Comment:
If find it worrysome that these all depend on offsets into the buffer that are only known at runtime. Do you think we really gain so much by avoiding the gaps from padding for dims that do not exists (which we had in the old implementation)?

Consider `increment()`, which (unless there is special "unrolled" handling in `transform`) represents the inner loop. It calls (among others), `coord(0)`, which is now computing:
```cpp
m_buffer.get() + (std::max(m_ndim, scipp::index{2}) * N + dim);
```
I am not worried about the math itself, but:
- We need 1 pointer (register) for the buffer.
- 1 register for `m_ndim`, which is only known at runtime

The old solution is simply `m_coord[0]`, which has a known offset from the current class instance at compile time, which can probably be encoded as an offset in the instructions.

While your benchmarks do not show major issues for common cases, I would tread carefully here. The pressure on the integer/index part of the CPU appears to be already very high (this is the reason why we had to do #1486 and #1539). Adding 2 more registers/addresses that need to be kept track of might make things worse for cases we have not considered yet.

Have you tried benchmarking cases that cannot rely on the strides optimization (maybe just disable those optimizations temporarily)? Note that even if that raises to red flags I would still be very hesitant to adopt the approach of this unified buffer, unless there is proven benefit.